x/stake/keeper_test.go

```diff
@@ -259,36 +266,178 @@ func TestGetValidators(t *testing.T) {
 	assert.Equal(t, candidates[3].Address, validators[1].Address, "%v", validators)
 }
 
-// TODO
 // test the mechanism which keeps track of a validator set change
 func TestGetAccUpdateValidators(t *testing.T) {
-	//TODO
+	ctx, _, keeper := createTestInput(t, nil, false, 0)
+
+	amts := []int64{9, 8, 7, 10, 6}
+	var candidatesIn [5]Candidate
+	for i, amt := range amts {
+		candidatesIn[i] = Candidate{
+			Address:     addrVals[i],
+			PubKey:      pks[i],
+			Assets:      sdk.NewRat(amt),
+			Liabilities: sdk.NewRat(amt),
+		}
+	}
+
+	validatorsEqual := func(t *testing.T, expected []Validator, actual []Validator) {
+		require.Equal(t, len(expected), len(actual))
+		for i := 0; i < len(expected); i++ {
+			assert.Equal(t, expected[i], actual[i])
+		}
+	}
+
+	genValidators := func(candidates []Candidate) []Validator {
+		validators := make([]Validator, len(candidates))
+		for i, c := range candidates {
+			validators[i] = c.validator()
+		}
+		return validators
+	}
+
+	candidates := []Candidate{candidatesIn[1], candidatesIn[4]}
```

Comment:
okay so I added some function named candidatesIN but this is really confusing generating the candidate group this way - ideally we don't simple "compile" the list as we've done here but retrieve the list from the keeper using "keeper.GetCandidates"